src/test/java/com/liferay/faces/util/cache/CacheTest.java

```diff
@@ -0,0 +1,319 @@
+		}
+	}
+
+	private void testCache(Cache cache, int iterations) {
+
+		for (int i = 0; i < iterations; i++) {
+
+			String key = "key" + i;
+			String value = "value" + i;
+			Assert.assertNull(cache.get(key));
+			Assert.assertEquals(value, cache.put(key, value));
+			Assert.assertEquals(value, cache.putIfAbsent(key, "different" + value));
+			Assert.assertEquals(value, cache.get(key));
+			Assert.assertEquals("different" + value, cache.put(key, "different" + value));
+			Assert.assertEquals("different" + value, cache.get(key));
+
+			if ((i % 2) == 1) {
+
+				Assert.assertEquals("different" + value, cache.remove(key));
+				Assert.assertNull(cache.get(key));
+			}
+		}
+
+		Assert.assertEquals(iterations / 2, cache.keySet().size());
+	}
+
+	private void testConcurrentCache(Cache<String, String> cache, Queue<Throwable> testErrors,
+		Queue<AssertionError> testFailures) throws Exception {
+
+		Thread[] threads = new Thread[100];
+		Random random = new Random(System.currentTimeMillis());
+
+		for (int i = 0; i < threads.length; i++) {
+			threads[i] = new TestConcurrentCacheThread(cache, testFailures, testErrors, "key" + (i % 25), "value" + i,
+					random.nextInt(25));
+		}
+
+		for (Thread thread : threads) {
+			thread.start();
+		}
+
+		errorOrFailTestIfNecessary(testErrors, testFailures);
+	}
+
+	private void testConcurrentMaxCacheCapacityCache(final Cache cache, final int maxCacheCapacity,
+		Queue<Throwable> testErrors, Queue<AssertionError> testFailures) throws Exception {
+
+		if (maxCacheCapacity < 10) {
+			throw new IllegalArgumentException("This test must be run with at least 10 values.");
+		}
+
+		new TestThreadBase(cache, testFailures, testErrors) {
+
+				@Override
+				protected void testCache() {
+
+					for (int i = 0; i < maxCacheCapacity; i++) {
+						cache.putIfAbsent("key" + i, "value" + i);
+					}
+				}
+			}.run();
+
+		new TestThreadBase(cache, testFailures, testErrors) {
+
+				@Override
+				protected void testCache() {
+					Assert.assertNotNull(cache.get("key0"));
+				}
+			}.run();
+
+		new TestThreadBase(cache, testFailures, testErrors) {
+
+				@Override
+				protected void testCache() {
+					cache.putIfAbsent("key1", "value1");
+				}
+			}.run();
+
+		new TestThreadBase(cache, testFailures, testErrors) {
+
+				@Override
+				protected void testCache() {
+					cache.put("key2", "value2");
+				}
+			}.run();
+
+		new TestThreadBase(cache, testFailures, testErrors) {
+
+				@Override
+				protected void testCache() {
+					cache.put("key" + maxCacheCapacity, "value" + maxCacheCapacity);
+				}
+			}.run();
+
+		new TestThreadBase(cache, testFailures, testErrors) {
+
+				@Override
+				protected void testCache() {
+					Assert.assertNull(cache.get("key3"));
+				}
+			}.run();
+
+		errorOrFailTestIfNecessary(testErrors, testFailures);
+	}
+
+	private abstract static class TestThreadBase extends Thread {
+
+		// Protected Final Data Members
+		protected final Cache<String, String> cache;
```

Comment:
@stiemannkj1 IDE warns that these "protected" data members should be package-private.